### DISCUSSION TITLE
Bevy Pointer Abstraction

### DISCUSSION THREAD

**Discussion by aevyrie** [bevyengine/bevy#5540] [Ideas]
Creating this as a central place for discussion of upstreaming this work into bevy: https://github.com/aevyrie/bevy_mod_picking/pull/137

The goal of this work is

1. Create a pointer abstraction usable by any kind of input, including mouse, touch, and software.
2. Make pointer interaction testable and mockable
3. Make the abstraction completely decoupled from the picking backend, whether it uses rapier, shader picking, or something completely custom.
4. Add pointer events - like enter, leave, drag, and drop - to any interactable entity, regardless of whether it's 2d UI or a 3D mesh.
5. Add event listeners with bubbling to make it super easy to express features like "if this entity is dragged onto another entity, delete it"

**Comment by Weibye** [comment]
I've looked at the code and it is IMHO very good. This is also one of the first crates I add when starting a new bevy project.

Any time the user wants to interact with entities in the world, this functionality is needed. We _are_ going to need these abstractions for the eventual Bevy Editor. This is also a hard requirement for full VR / XR support. 

The way I see it is we either push to upstreaming it soon, or we upstream it later, but I don't think _not upstreaming_ it at all is a viable option. 

The benefit of doing it sooner rather than later is that can be battle tested for longer and drive further improvements to other parts of the engine. It would allow for more rapid and smaller iteration cycles to improve the system over time.
